script: add ram and rom total size to json file

wangnuannuan opened the pull request:
We want to get the rom and ram size by reading the JSON file. Because it's safer and easier to parse JSON file rather than extract total size from `stdout`. But the printed total size differs from calculated from JSON file https://github.com/zephyrproject-rtos/zephyr/issues/33967.

Take rom size as example. The generated `rom.json' contains only symbols data,  I think we can convert the target data as a python dict which including 2 key-value pairs, and then save this target data into JSON file.
```
            target = dict()
            target["symbols"] = json.loads(data) # The symbols data stored in the JSON file before
            target["total_size"] = symsize
```



Signed-off-by: Jingru Wang <jingru@synopsys.com>

abrodkin reviewed (changes requested):
To me we have 2 separate things here to discuss:
1. Possibly wrong calculation of the total size the way we do it now. That needs to be understood and fixed if necessary.
2. Motivation to move total size into the .json file: IMHO it's much easier to extract that value from .json file as compared to doing the math. As for currently printed value I'd say it's safer to parse a .json (knowing exact element name) rather than extracting something from `stdout`. What if we decide to make the printed number fancier and convert it to say kilobytes?

Also it would be good if there's a snippet of updated .json file so that we may see how it looks like with new data field inside.

That said the following needs to be done:
1. Understand why currently printed total size differs from calculated from .json
2. Re-word motivation for the change
3. Add a snippet of the new .json to the PR description

abrodkin commented:
@wangnuannuan could you please take a look at my comments above?

wangnuannuan committed:
script: add ram and rom total size to json file

It's easier and safer to parse json file rather than extract
total size from stdout

Now the .json file has two keys:
{
    "symbols": {
       "children": [
           ...
	],
       "identifier": ":",
       "name": "root",
       "size": 2220
    },
    "total_size": 34272
}

Signed-off-by: Jingru Wang <jingru@synopsys.com>

abrodkin commented:
@wangnuannuan could you please also add an example (or just a snippet) of generated `.json` so that we may see how it's going to look like after your change?

wangnuannuan commented:
Take `platform=iotdk zephyr-version=zephyr-v2.5.0-2384-gfacd0f5  target=rom` as an example. The previous rom.json:
```
{
  "children": [
      ...
  ],
  "identifier": ":",
  "name": "root",
  "size": 2220
}
```
Now the rom.json
```
{
    "symbols": {
        "children": [
            ...
        ],
        "identifier": ":",
        "name": "root",
        "size": 2220
    },
    "total_size": 34292
}
```

tejlmand commented on scripts/footprint/size_report:
```diff
@@ -635,8 +636,11 @@ def main():
             exporter = JsonExporter(indent=2, sort_keys=True)
             data = exporter.export(root)
+            target = dict()
+            target["symbols"] = json.loads(data)
```
Why not change the exporter in the line above to be a dict exporter ?

That way you don't need to export to json and load into a dict, just to reformat back to json  at line 643 when doing `json.dump`.

Basically, what i'm proposing is:
```
            exporter = DictExporter()
            data = dict()
            data["symbols"] = exporter.export(root)
            data["total_size"] = symsize
            with open(args.json, "w") as fp:
                json.dump(data, fp, indent=4)
```

In case you do this change, remember to update line 33 to:
```
from anytree.exporter import DictExporter
```

wangnuannuan replied:
OK, thank you for your suggestion.

tejlmand reviewed (commented):
Minor proposal.

tejlmand reviewed (approved):
lgtm, thanks.

nashif reviewed (changes requested):
This is breaking compatibility with anytree, using this simple script I can load the json file into anytree and print it:

```
#!/usr/bin/env python3

from anytree.importer import JsonImporter

from anytree import RenderTree, Node, RenderTree, AsciiStyle, PreOrderIter
import sys

importer = JsonImporter()

with open(sys.argv[1], "r") as f:
    data = f.read()
    root1 = importer.import_(data)
    print(RenderTree(root1, style=AsciiStyle()))
```

output:

```
python3 im.py build/ram.json
AnyNode(identifier=':', name='root', size=3088)
|-- AnyNode(identifier='drivers', name='drivers', size=156)
|   |-- AnyNode(identifier='drivers/console', name='console', size=4)
|   |   +-- AnyNode(identifier='drivers/console/uart_console.c', name='uart_console.c', size=4)
|   |       +-- AnyNode(identifier='drivers/console/uart_console.c/uart_console_dev', name='uart_console_dev', size=4)
|   |-- AnyNode(identifier='drivers/timer', name='timer', size=20)
|   |   |-- AnyNode(identifier='drivers/timer/cortex_m_systick.c', name='cortex_m_systick.c', size=16)
|   |   |   |-- AnyNode(identifier='drivers/timer/cortex_m_systick.c/announced_cycles', name='announced_cycles', size=4)
|   |   |   |-- AnyNode(identifier='drivers/timer/cortex_m_systick.c/cycle_count', name='cycle_count', size=4)
|   |   |   |-- AnyNode(identifier='drivers/timer/cortex_m_systick.c/last_load', name='last_load', size=4)
|   |   |   +-- AnyNode(identifier='drivers/timer/cortex_m_systick.c/overflow_cyc', name='overflow_cyc', size=4)
|   |   +-- AnyNode(identifier='drivers/timer/sys_clock_init.c', name='sys_clock_init.c', size=4)
|   |       +-- AnyNode(identifier='drivers/timer/sys_clock_init.c/__devstate_sys_init_sys_clock_driver_init0', name='__devstate_sys_init_sys_clock_driver_init0', size=4)
..
...
```

This is no longer possible with this change.

tejlmand commented:
@nashif but you can do:
```
#!/usr/bin/env python3

from anytree.importer import DictImporter

from anytree import RenderTree, Node, RenderTree, AsciiStyle, PreOrderIter
import sys
import json

importer = DictImporter()

with open(sys.argv[1], "r") as f:
    data = json.load(f)
    root1 = importer.import_(data['symbols'])
    print(RenderTree(root1, style=AsciiStyle()))
```

Isn't that acceptable ?

Output:
```
>>> print(RenderTree(root1, style=AsciiStyle()))
AnyNode(identifier=':', name='root', size=568)
|-- AnyNode(identifier='lib', name='lib', size=3194)
|   |-- AnyNode(identifier='lib/os', name='os', size=2936)
|   |   |-- AnyNode(identifier='lib/os/printk.c', name='printk.c', size=94)
|   |   |   |-- AnyNode(identifier='lib/os/printk.c/char_out', name='char_out', size=16)
|   |   |   |-- AnyNode(identifier='lib/os/printk.c/_char_out', name='_char_out', size=4)
|   |   |   |-- AnyNode(identifier='lib/os/printk.c/__printk_hook_install', name='__printk_hook_install', size=12)
|   |   |   |-- AnyNode(identifier='lib/os/printk.c/arch_printk_char_out', name='arch_printk_char_out', size=4)
|   |   |   |-- AnyNode(identifier='lib/os/printk.c/vprintk', name='vprintk', size=32)
|   |   |   +-- AnyNode(identifier='lib/os/printk.c/printk', name='printk', size=26)
|   |   |-- AnyNode(identifier='lib/os/onoff.c', name='onoff.c', size=786)
....
```

nashif commented:
Ok, fair enough
